cpp/src/arrow/io/buffered.cc

```diff
@@ -305,10 +305,13 @@ class BufferedInputStream::Impl : public BufferedBase {
       RETURN_NOT_OK(BufferIfNeeded());
     }
 
-    // Increase the buffer size if needed
+    // Increase the buffer size if needed.
+    // SetBufferSize(...) will reset buffer_pos_, so backup buffer_pos_ and restore it.
     if (nbytes > buffer_->size() - buffer_pos_) {
+      int64_t original_buffer_pos = buffer_pos_;
       RETURN_NOT_OK(SetBufferSize(nbytes + buffer_pos_));
       DCHECK(buffer_->size() - buffer_pos_ >= nbytes);
+      buffer_pos_ = original_buffer_pos;
```

Comment:
I think a better fix may be to not modify `buffer_pos_` at all in `SetBufferSize`. I think this will also have implications for modifying the buffer size in a BufferedOutputStream that has data buffered. I'll take a closer look